### PR TITLE
Type42 subsetting in PS/PDF

### PR DESCRIPTION
**Pull request by aitikgupta**
## PR Summary
This PR is a fresh rebase from https://github.com/matplotlib/matplotlib/pull/18143.

- Adds a dependency: [`fonttools`](https://github.com/fonttools/fonttools) to handle font subsetting for us
  (we already have an external ttconv dependency, which does not handle subsetting)
- Interfaces a `getSubset` utility to get file-like objects containing subsetted font data

Possibly fixes https://github.com/matplotlib/matplotlib/issues/11303 (large file sizes)
Fixes https://github.com/matplotlib/matplotlib/issues/18191.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by aitikgupta** [comment]
(just trying, since this is a draft)
@meeseeksmachine hello

**Comment by aitikgupta** [comment]
(just trying, since this is a draft)
@meeseeksdev hello

**Comment by aitikgupta** [comment]
The last commit subsets and embeds Type 42 PS/EPS, here's the size difference in outputs:

```
nosub-dejavu.ps                                    ----> 1.2 MB
sub-dejavu.ps                                      ----> 8.5 kB
```

(To test it out on your machine, apply this patch for `nosub-dejavu.ps`:)
```diff
diff --git a/lib/matplotlib/backends/backend_ps.py b/lib/matplotlib/backends/backend_ps.py
index 2a3ab64a1c..ed0af4ea85 100644
--- a/lib/matplotlib/backends/backend_ps.py
+++ b/lib/matplotlib/backends/backend_ps.py
@@ -997,12 +997,8 @@ class FigureCanvasPS(FigureCanvasBase):
                             ) as tmp:
                                 tmp.write(fontdata)
                                 tmp.seek(0, 0)
-                                font = FT2Font(tmp.name)
-                                glyph_ids = [
-                                    font.get_char_index(c) for c in chars
-                                ]
                                 convert_ttf_to_ps(
-                                    os.fsencode(tmp.name),
+                                    os.fsencode(font_path),
                                     fh,
                                     fonttype,
                                     glyph_ids,
```

The script:
```python
import matplotlib.pyplot as plt

plt.rcParams["ps.fonttype"] = 42
plt.figtext(0.5, 0.5, "hello")
# plt.savefig('sub-dejavu.ps')                    # before applying the patch
# plt.savefig('nosub-dejavu.ps')                  # after applying the patch
```

File: lib/matplotlib/backends/backend_pdf.py
Comment (anntzer) [inline lib/matplotlib/backends/backend_pdf.py:1214]: FT2Font can handle file objects too (things that have a read()), so you don't need to go through the filesystem; a BytesIO should be good enough.

File: lib/matplotlib/backends/backend_pdf.py
Comment (anntzer) [inline lib/matplotlib/backends/backend_pdf.py:1199]: turn these into logs?

File: lib/matplotlib/backends/backend_ps.py
Comment (anntzer) [inline lib/matplotlib/backends/backend_ps.py:997]: as above re: intermediate file.

File: lib/matplotlib/backends/backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/backends/backend_pdf.py:1214]: I see, this can be done for PDF backend (however, see note on PS backend)

File: lib/matplotlib/backends/backend_ps.py
Comment (aitikgupta) [reply lib/matplotlib/backends/backend_ps.py:997]: `convert_ttf_to_ps` takes a filepath as the first parameter, because read_font takes fontpath too:
https://github.com/matplotlib/matplotlib/blob/d1567715f1913ce935462c408f02e70e6299b6cb/extern/ttconv/pprdrv_tt.cpp#L1223

for PDF backend we're not using this function, but here I think we need to? (unless there's a way around it)

File: lib/matplotlib/backends/backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/backends/backend_pdf.py:1199]: yeah, was waiting for a general review before changing the previous PR's contents

File: lib/matplotlib/backends/backend_ps.py
Comment (anntzer) [reply lib/matplotlib/backends/backend_ps.py:997]: I guess it may be worth seeing whether ttconv can be fixed to support filelikes as well, but this can be done later.

Comment (aitikgupta) [reply lib/matplotlib/backends/backend_ps.py:997]: `read_font` it's being called from the below function (which is called by `convert_ttf_to_ps`): https://github.com/matplotlib/matplotlib/blob/d1567715f1913ce935462c408f02e70e6299b6cb/src/_ttconv.cpp#L132

Comment (aitikgupta) [reply lib/matplotlib/backends/backend_ps.py:973]: > I guess it may be worth seeing whether ttconv can be fixed to support filelikes as well, but this can be done later.

I added a TODO just before it

**Comment by aitikgupta** [comment]
How do we add dependencies for CI?
(it's failing because it's unable to find `fonttools`)

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (QuLogic) [inline lib/matplotlib/backends/_backend_pdf_ps.py:22]: This is not a valid PEP8 name. Also, it's not a descriptive name; `matplotlib.backends._backend_pdf_ps.getSubset` does not really indicate what you're subsetting.

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (QuLogic) [inline lib/matplotlib/backends/_backend_pdf_ps.py:27]: It's not public, but you should still try to use numpydoc.

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (QuLogic) [inline lib/matplotlib/backends/_backend_pdf_ps.py:41]: Can this be used as a context manager instead?

File: lib/matplotlib/backends/backend_pdf.py
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:1198]: Logging should usually use `%`.

File: lib/matplotlib/backends/backend_ps.py
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py:960]: This appears to require testing.

File: setup.py
Comment (QuLogic) [inline setup.py:328]: This minimum version needs to go in `requirements/testing/minver.txt` so it's actually tested.

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (QuLogic) [inline lib/matplotlib/backends/_backend_pdf_ps.py:31]: Comment what `FFTM` is.

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (aitikgupta) [reply lib/matplotlib/backends/_backend_pdf_ps.py]: I changed the name to `get_glyphs_subset`

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (aitikgupta) [reply lib/matplotlib/backends/_backend_pdf_ps.py:27]: I added `Parameters` section, if that's what you meant?

**Comment by aitikgupta** [comment]
> https://github.com/matplotlib/matplotlib/pull/20391#discussion_r653337251: This appears to require testing.

I think it's difficult to test `get_glyphs_subset`, since it really depends on font files, and if we chose certain characters to test with a fixed font (lets say), there's no guarantee that the subset will be the same.

Is there any other way to test this?

File: setup.py
Comment (aitikgupta) [reply setup.py:328]: hmm, the tests still fail with `NoModuleFound` error even after adding it to minver.txt
Any idea?

**Comment by jklymak** [comment]
You can mark as ready-to-review at any point.  I just kick things to Draft so they don't show up in the queue if the PR requires action on the part of the author.  (this one is still not passing the tests).

**Comment by aitikgupta** [comment]
> (this one is still not passing the tests).

Yeah, CI isn't installing fonttools, which is why `NoModuleFound` is all over the place (even after adding it to `minvers.txt`: https://github.com/matplotlib/matplotlib/pull/20391#discussion_r654688358)

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (QuLogic) [reply lib/matplotlib/backends/_backend_pdf_ps.py:27]: Yes, that will work.

**Comment by QuLogic** [comment]
Install is done with `--no-deps`, so hard dependencies also need to be listed here: https://github.com/matplotlib/matplotlib/blob/master/.github/workflows/tests.yml#L148 `minver.txt` is only a _constraint_ on versions.

**Comment by QuLogic** [comment]
> I think it's difficult to test `get_glyphs_subset`, since it really depends on font files, and if we chose certain characters to test with a fixed font (lets say), there's no guarantee that the subset will be the same for different versions of the library.

We ship DejaVu ourselves, so that should always be available for testing.
 
> Is there any other way to test this?

A file with just 'A' embedded should be smaller than one with the full alphabet, say?
Also, a subsetted PDF should appear the same as one without subsetting (assuming Ghostscript doesn't happen to substitute the same glyphs.)

**Comment by aitikgupta** [comment]
> A file with just 'A' embedded should be smaller than one with the full alphabet, say?

Oh, if we just test the subset being 'smaller' yeah we can do it, but not with the exact number glyphs, since that will definitely vary. (we also set `recommeded_glyphs`=True)

But even so, isn't that the 'full-time job' of the fonttools library itself? or in other words wouldn't testing the `get_glyphs_subset` function be the same as testing the library itself (that it does reduce the number of glyphs)?

File: lib/matplotlib/backends/backend_pdf.py
Comment (anntzer) [inline lib/matplotlib/backends/backend_pdf.py:1205]: probably better to stick to ascii in the logs just in case someone is printing these to an old-style windows cmd.exe which may not like that.

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (jkseppan) [inline lib/matplotlib/backends/_backend_pdf_ps.py:22]: I suggest adding a unit test for this function: call the function with some font and some characters, read the output using FT2Font, check that the characters you want are in there and (some) characters you don't want are not in there.

File: lib/matplotlib/backends/backend_ps.py
Comment (anntzer) [inline lib/matplotlib/backends/backend_ps.py:981]: I don't think you need the seek() here.

**Comment by jkseppan** [comment]
Oh, and one more thing (sorry for not remembering this earlier): the [PDF specification](https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf) has some extra requirements for font subsets in section 9.6.4. The PostScript name of subsetted fonts needs to have a prepended tag of six random uppercase letters and a plus sign, e.g. `EOODIA+Poetica` if the original font name is `Poetica`. This is relevant in the case that multiple Matplotlib plots are combined in the same document, e.g. in a LaTeX paper with several figures. The random tags prevent collisions between the different versions of the same font.

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (aitikgupta) [reply lib/matplotlib/backends/_backend_pdf_ps.py:22]: Added a bunch of tests (not sure about what we _don't_ want in there though

**Comment by aitikgupta** [comment]
> extra requirements for font subsets ... PostScript name of subsetted fonts needs to have a prepended tag of six random uppercase letters and a plus sign

I think we don't do this even for Type 3 subsetting?
Since we _always_ try to subset those, I can probably just add a function to modify the name for Type 3 and Type 42 both.

**Comment by aitikgupta** [comment]
This breaks the `test_determinism_check`:
https://github.com/matplotlib/matplotlib/blob/593de3550e578aed1fc2e392041bb7796fcc3ce1/lib/matplotlib/tests/test_determinism.py#L85-L88

This is for obvious reasons, since we have random string inside font postscript name. (due to the specification)

**Comment by jkseppan** [comment]
Then the string shouldn't be random, but perhaps something derived from the subset of glyphs? Take `hash(frozenset(glyphs))`, convert it to base 26 and take the first six letters, or something like that.

The PDF specification only requires the tag for Type 1 and TrueType fonts. It probably doesn't hurt with Type 3 fonts either.

**Comment by aitikgupta** [comment]
> Then the string shouldn't be random, but perhaps something derived from the subset of glyphs?

That totally makes sense! Let me try this out 👍🏼

File: lib/matplotlib/backends/_backend_pdf_ps.py
Comment (jkseppan) [inline lib/matplotlib/backends/_backend_pdf_ps.py:19]: Someone could see it as bad behaviour from a library if it modifies the logging configuration for another library. If this is about the unknown truetype table, I think it's fine to have our own list of dropped tables and add to it whenever someone complains about messages with a new font.

Comment (aitikgupta) [reply lib/matplotlib/backends/_backend_pdf_ps.py:19]: That makes sense.
I'll fix this in the next commit

File: lib/matplotlib/backends/backend_ps.py
Comment (sauerburger) [inline lib/matplotlib/backends/backend_ps.py:980]: I think we should add a `tmp.flush()` to avoid a race condition between the content being written out and `convert_ttf_to_ps()` reading it back.

Comment (aitikgupta) [reply lib/matplotlib/backends/backend_ps.py:980]: I was under the impression that Python interpreter will only move to the next line once the `write` is complete?

Seems like `.flush()` 'clears out the internal buffer of the file', which is still in memory here, but once we've written it out on disk, could a race condition still exist?

(asking totally on the learning side here, I can add this in my next commit)

Comment (sauerburger) [reply lib/matplotlib/backends/backend_ps.py:980]: Write operations on a file object in Python are usually [buffered](https://docs.python.org/3/library/functions.html#open) with a [default buffer size](https://docs.python.org/3/library/io.html#io.DEFAULT_BUFFER_SIZE) of 4KiB or 8KiB. Calling `flush()` ensures that another process can read back the content, [see](https://stackoverflow.com/questions/7127075/what-exactly-is-pythons-file-flush-doing). The OS might still buffer and delay the write operation to the physical disk for performance reasons. This is completely transparent to user programs. To ensure that the data is permanently written to disk and persisted in case of a system crash, you could call `fsync()` but this is not necessary in this case.

Comment (aitikgupta) [reply lib/matplotlib/backends/backend_ps.py:980]: Thank you so much for a detailed answer! (that SO link was an amazing and insightful read)

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (jkseppan) [inline lib/matplotlib/tests/test_backend_pdf.py:17]: I think this is an assertion that a warning is raised, which will probably break when a future version of fonttools fixes the code that causes the warning. A suitable [filterwarnings setting](https://docs.pytest.org/en/6.2.x/warnings.html#deprecationwarning-and-pendingdeprecationwarning) in pytest.ini would let us ignore fonttools specific warnings.

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (jkseppan) [inline lib/matplotlib/tests/test_backend_pdf.py:370]: Shouldn't we require strict inequality here? We specifically want to exclude a lot of glyphs.

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (jkseppan) [inline lib/matplotlib/tests/test_backend_pdf.py:367]: This looks to me like the comment is contradicting the code. `set(chars)` is the set of characters we want to be available, so it should be a subset (not superset) of the set of characters available in the reduced font.

Comment (jkseppan) [reply lib/matplotlib/tests/test_backend_pdf.py:367]: ... but since the test passes, I wonder what's going on. Could you check what the sets on both sides of the `.issuperset` operation actually are? I wonder if there's a bug that makes the right-hand set be empty.

**Comment by jkseppan** [comment]
I think this is getting close to mergeable. Changes requested are the `.flush` call, the `filterwarnings` thing, and some minor points about the added test. This should also be documented in `doc/users/next_whats_new/`.

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/tests/test_backend_pdf.py:17]: Even after adding this to `pytest.ini`, this failed.
However, adding it to `lib/matplotlib/testing/conftest/py` worked.

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/tests/test_backend_pdf.py:370]: Testing with font files that are _already_ subsetted with certain characters, the strict equality would fail.

Or for fonts that cannot be subsetted at all.

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/tests/test_backend_pdf.py:367]: Although the left and the right sides are valid (they print the set of characters):
```console
{'o', 'b', 'd', 's', ' ', '!', 'u', 'h', 'e', 't', 'l'}
```
It seems LHS is equal to RHS, which is why the tests pass. I agree it should be `.subset()`. (though I'm not able to recall why I put superset in the first place :/)

**Comment by aitikgupta** [comment]
Bringing up what @jkseppan said on gitter a while ago:
> We should probably try to default to Type 42 and fix any problems our Type 42 conversion has, because that's what publishers expect and it's probably more useful for other tools that consume pdf files.

Should we consider changing matplotlib's default fonttype from `3` to `42`?

**Comment by sauerburger** [comment]
Over the last few days I've submitted a couple of issues (https://github.com/matplotlib/matplotlib/issues/20628, https://github.com/matplotlib/matplotlib/issues/20616, https://github.com/matplotlib/matplotlib/issues/20614, https://github.com/matplotlib/matplotlib/issues/20612) and corresponding PRs, all addressing specific issues I've encountered with Type 42 fonts. I just wanted to give a heads up, that this PR does not interfere with any of the other PRs (after one manual merge).

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/tests/test_backend_pdf.py:367]: ^^ change stuff according to above comment

Comment (jkseppan) [reply lib/matplotlib/tests/test_backend_pdf.py:367]: The DejavuSerif.ttf we distribute is not subsetted, so it should get much smaller.

**Comment by aitikgupta** [comment]
As discussed on dev call, we could add a test triggering the whole route for Type 42 subsetting in PostScript backend.

**Comment by aitikgupta** [comment]
^^ Also check out PS /fontname output

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (jkseppan) [inline lib/matplotlib/tests/test_backend_pdf.py:364]: This could be `set(chars) == set(chr(key) for ...)`

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (jkseppan) [inline lib/matplotlib/tests/test_backend_pdf.py:367]:
```suggestion
    assert len(subcmap) < len(nosubcmap)
```

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/tests/test_backend_pdf.py]: Done 👍🏼

File: lib/matplotlib/tests/test_backend_pdf.py
Comment (aitikgupta) [reply lib/matplotlib/tests/test_backend_pdf.py]: Fixed this 👍🏼

**Comment by aitikgupta** [comment]
^Rebased, fixed subset tests, added a new test for multiple fonttypes in PS backend.

**Comment by aitikgupta** [comment]
Seems like we can't do a `tempfile.NamedTemporaryFile` in tests, it raises a:
```console
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\tmpp1lzen9i.ps'
```
for Azure pipelines..

Can we otherwise get a buffer from a figure? (so instead of fig.savefig() we could use something like a  fig.get_buffer(), which can get the job done)

**Comment by anntzer** [comment]
`buf = io.BytesIO(); fig.savefig(buf, fmt="ps"); buf.getvalue()`?

**Comment by aitikgupta** [comment]
> buf = io.BytesIO(); fig.savefig(buf, fmt="ps"); buf.getvalue()?

It was right there in one of the tests too, I completely forgot about it. I'll push again..

**Comment by jkseppan** [comment]
pytest has a tmpdir fixture, perhaps it does whatever tricks are needed on Windows:  https://docs.pytest.org/en/6.2.x/tmpdir.html

It's being used in some tests (e.g. https://github.com/matplotlib/matplotlib/blob/c4dcf5e628d0441b21df1c8fe7cac4834a6944e9/lib/matplotlib/tests/test_mathtext.py#L371).

**Comment by aitikgupta** [comment]
> https://dev.azure.com/matplotlib/matplotlib/_build/results?buildId=20011&view=logs&j=a0e323ef-73e6-5069-91af-81f1eca08ea9&t=aaf81cb8-bb95-512e-5f32-9726d69a7874&l=152

`>     fig.savefig(buf, format="ps")`
this raises :(


```console
RuntimeError: Failed to open TrueType font
```


This is triggered by ttconv here:
> https://dev.azure.com/matplotlib/matplotlib/_build/results?buildId=20011&view=logs&j=a0e323ef-73e6-5069-91af-81f1eca08ea9&t=aaf81cb8-bb95-512e-5f32-9726d69a7874&l=241

`>    convert_ttf_to_ps(`


I don't get this locally, and also, all the tests pass except the azure pipeline.. I'm not sure how can I reproduce? 😅

**Comment by jkseppan** [comment]
It could be because of this: https://bugs.python.org/issue14243

Apparently on Windows you have to close the file before opening it for reading, and then clean it up manually.

**Comment by anntzer** [comment]
Haven't followed the whole thread so I'm not sure it's the problem at hand, but my usual strategy is to never (well, rarely) use NamedTemporaryFile exactly because of Windows support, and to always instead use a TemporaryDirectory and create a file with whatever name you want in it.

**Comment by aitikgupta** [comment]
That worked!
I wonder, if `tempfile.NamedTemporaryFile` is really that buggy (or doesn't exactly work on Windows in the same way), should it really be a part of Python's API?

Or from a different perspective, since `TemporaryDirectory` worked, it wouldn't be so hard to create a `NamedTemporaryFile` from it, right? But one works, the other doesn't..

**Comment by jkseppan** [comment]
> That worked!
> I wonder, if `tempfile.NamedTemporaryFile` is really that buggy (or doesn't exactly work on Windows in the same way), should it really be a part of Python's API?

As you can see from the discussion on that 9-year-old bug, reasonable people disagree on the solution. Or looking at some of the later messages, perhaps the issue needs someone to choose a solution, submit a PR and keep pushing until it is fixed. In any case, changing the API of anything in the Python standard library is a slow process, since someone somewhere is going to depend on the current behaviour.

> Or from a different perspective, since `TemporaryDirectory` worked, it wouldn't be so hard to create a `NamedTemporaryFile` from it, right? But one works, the other doesn't..

The current Windows implementation of `NamedTemporaryFile` guarantees that the file is deleted if the process exits abruptly, and I don't think the `TemporaryDirectory` solution can do that.

**Review by jkseppan — APPROVED** [review]
This is starting to look ready to me. Thanks for all the work!

**Comment by jkseppan** [comment]
I removed the "needs revision" label - I think some busy developers use them to organize their reviews, so removing the label is a signal to take another look.

**Review by tacaswell — APPROVED** [review]
Modulo adding dependency in dependency.rst as well.

**Comment by tacaswell** [comment]
It may also be worth at API change note that we added a new dependency?

File: doc/api/next_api_changes/development/20391-AG.rst
Comment (jkseppan) [inline doc/api/next_api_changes/development/20391-AG.rst:5]:
```suggestion
is now required for Maptlotlib 3.5.
```

Comment (jkseppan) [reply doc/api/next_api_changes/development/20391-AG.rst:5]: I think  the wording "is integrated in with" could be read to mean that we ship a vendored copy of the library.

**Comment by sauerburger** [comment]
I tested this PR with the current master (which contains a fix for characters with code point >65535, https://github.com/matplotlib/matplotlib/pull/20633). I think there is a small incompatibility. As far as I can tell the subsetted font does not contain chars with code point >65535. This is good, since we will embed them as XObjects anyway. However, when we embed them as XObjects, we currently take them from the subsetted font (that does not contain them anymore).

To fix it, we could replace `font` in https://github.com/matplotlib/matplotlib/blob/73b7abf14c77014ab2436e7691e19cbee5864f4b/lib/matplotlib/backends/backend_pdf.py#L1302 and https://github.com/matplotlib/matplotlib/blob/73b7abf14c77014ab2436e7691e19cbee5864f4b/lib/matplotlib/backends/backend_pdf.py#L1305 with a copy of the original font before subsetting, e.g.

```python
            full_font = font  # Ref to complete font, used for XObjects
            font = FT2Font(fontdata)
```

**Comment by jklymak** [comment]
@sauerburger can you suggest a test for that, or is it already in #20633?

**Comment by aitikgupta** [comment]
> can you suggest a test for that, or is it already in #20633?

The test is already in, and that is exactly what is breaking the CI for this PR.
I'll rebase and push a commit, fixing the error..

**Comment by jkseppan** [comment]
It seems to me that the requested changes have been made.

File: lib/matplotlib/backends/backend_ps.py
Comment (QuLogic) [reply lib/matplotlib/backends/backend_ps.py]: The context manager calls `.close`, which flushes the buffer; there's no need for an explicit `.flush` here (at least in how the code ended up as merged.)

**Pull request merged**
